File: core/src/main/java/vn/loitp/views/navigationview/LNavigationView.java
package vn.loitp.views.navigationview;

import android.content.Context;
import android.graphics.Color;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.daimajia.androidanimations.library.Techniques;

import java.util.List;

import loitp.core.R;
import vn.loitp.core.utilities.LAnimationUtil;

/**
 * Created by www.muathu@gmail.com on 5/13/2017.
 */

public class LNavigationView extends RelativeLayout implements View.OnClickListener {
    private final String TAG = getClass().getSimpleName();
    private ImageView ivPrev;
    private ImageView ivNext;
    private TextView tv;
    private List<String> stringList;
    private int currenIndex;
    private int colorOn = Color.BLACK;
    private int colorOff = Color.GRAY;
    private boolean isEnableAnimation = true;

    private NVCallback nvCallback;

    public interface NVCallback {
        public void onIndexChange(int index, String s);
    }

    public void setNVCallback(NVCallback nvCallback) {
        this.nvCallback = nvCallback;
    }

    public LNavigationView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public LNavigationView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    private void init() {
        inflate(getContext(), R.layout.view_navigation, this);
        this.ivPrev = (ImageView) findViewById(R.id.iv_prev);
        this.ivNext = (ImageView) findViewById(R.id.iv_next);
        this.tv = (TextView) findViewById(R.id.tv);
        ivPrev.setOnClickListener(this);
        ivNext.setOnClickListener(this);
        setEnableController(ivPrev, false);
        setEnableController(ivNext, false);
    }

    private void setEnableController(ImageView imageView, boolean isEnable) {
        if (isEnable) {
            imageView.setEnabled(true);
            imageView.setClickable(true);
            imageView.setColorFilter(colorOn);
        } else {
            imageView.setEnabled(false);
            imageView.setClickable(false);
            imageView.setColorFilter(colorOff);
        }
    }

    public void setStringList(List<String> stringList) {
        if (stringList == null || stringList.isEmpty()) {
            return;
        }
        currenIndex = 0;
        this.stringList = stringList;
        updateUI();
    }

    public void setCurrenIndex(int index) {
        if (stringList == null || stringList.isEmpty()) {
            return;
        }
        if (index < 0 || index > stringList.size() - 1) {
            return;
        }
        currenIndex = index;
        updateUI();
    }

    private void updateUI() {
        String s = stringList.get(currenIndex);
        tv.setText(s);
        if (isEnableAnimation) {
            LAnimationUtil.play(tv, Techniques.FlipInX);
        }
        int size = stringList.size();
        if (size == 1) {
            setEnableController(ivPrev, false);
            setEnableController(ivNext, false);
        } else {
            if (currenIndex <= 0) {
                setEnableController(ivPrev, false);
                setEnableController(ivNext, true);
            } else if (currenIndex >= size - 1) {
                setEnableController(ivPrev, true);
                setEnableController(ivNext, false);
            } else {
                setEnableController(ivPrev, true);
                setEnableController(ivNext, true);
            }
        }
        if (nvCallback != null) {
            nvCallback.onIndexChange(currenIndex, s);
        }
    }

    @Override
    public void onClick(View view) {
        if (view == ivPrev) {
            if (isEnableAnimation) {
                LAnimationUtil.play(view, Techniques.Pulse);
            }
            currenIndex--;
            updateUI();
        } else if (view == ivNext) {
            if (isEnableAnimation) {
                LAnimationUtil.play(view, Techniques.Pulse);
            }
            currenIndex++;
            updateUI();
        }
    }

    public ImageView getIvPrev() {
        return ivPrev;
    }

    public ImageView getIvNext() {
        return ivNext;
    }

    public TextView getTv() {
        return tv;
    }

    public List<String> getStringList() {
        return stringList;
    }

    public int getCurrenIndex() {
        return currenIndex;
    }

    public int getColorOn() {
        return colorOn;
    }

    public void setColorOn(int colorOn) {
        this.colorOn = colorOn;
    }

    public int getColorOff() {
        return colorOff;
    }

    public void setColorOff(int colorOff) {
        this.colorOff = colorOff;
    }

    public boolean isEnableAnimation() {
        return isEnableAnimation;
    }

    public void setEnableAnimation(boolean enableAnimation) {
        isEnableAnimation = enableAnimation;
    }
}

File: core/src/main/java/com/views/navigationview/LNavigationView.kt
package com.views.navigationview

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.View
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import com.R
import com.core.utilities.LAnimationUtil
import com.daimajia.androidanimations.library.Techniques

class LNavigationView : RelativeLayout, View.OnClickListener {
    var ivPrev: ImageView? = null
    var ivNext: ImageView? = null
    var tv: TextView? = null

    private var stringList = ArrayList<String>()
    private var currenIndex = 0
    var colorOn = Color.BLACK
    var colorOff = Color.GRAY
    var isEnableAnimation = true

    private var nvCallback: NVCallback? = null

    interface NVCallback {
        fun onIndexChange(index: Int, s: String?)
    }

    fun setNVCallback(nvCallback: NVCallback?) {
        this.nvCallback = nvCallback
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle) {
        init()
    }

    private fun init() {
        View.inflate(context, R.layout.view_navigation, this)

        ivPrev = findViewById(R.id.ivPrev)
        ivNext = findViewById(R.id.ivNext)
        tv = findViewById(R.id.tv)

        ivPrev?.setOnClickListener(this)
        ivNext?.setOnClickListener(this)

        setEnableController(imageView = ivPrev, isEnable = false)
        setEnableController(imageView = ivNext, isEnable = false)
    }

    private fun setEnableController(imageView: ImageView?, isEnable: Boolean) {
        if (isEnable) {
            imageView?.let {
                it.isEnabled = true
                it.isClickable = true
                it.setColorFilter(colorOn)
            }
        } else {
            imageView?.let {
                it.isEnabled = false
                it.isClickable = false
                it.setColorFilter(colorOff)
            }
        }
    }

    fun setStringList(stringList: ArrayList<String>?) {
        if (stringList == null || stringList.isEmpty()) {
            return
        }
        currenIndex = 0
        this.stringList = stringList
        updateUI()
    }

    fun setCurrenIndex(index: Int) {
        if (stringList.isEmpty()) {
            return
        }
        if (index < 0 || index > stringList.size - 1) {
            return
        }
        currenIndex = index
        updateUI()
    }

    private fun updateUI() {
        val s = stringList[currenIndex]
        tv?.text = s
        if (isEnableAnimation) {
            LAnimationUtil.play(view = tv, techniques = Techniques.FlipInX)
        }
        val size = stringList.size
        if (size == 1) {
            setEnableController(imageView = ivPrev, isEnable = false)
            setEnableController(imageView = ivNext, isEnable = false)
        } else {
            when {
                currenIndex <= 0 -> {
                    setEnableController(imageView = ivPrev, isEnable = false)
                    setEnableController(imageView = ivNext, isEnable = true)
                }
                currenIndex >= size - 1 -> {
                    setEnableController(imageView = ivPrev, isEnable = true)
                    setEnableController(imageView = ivNext, isEnable = false)
                }
                else -> {
                    setEnableController(imageView = ivPrev, isEnable = true)
                    setEnableController(imageView = ivNext, isEnable = true)
                }
            }
        }
        nvCallback?.onIndexChange(index = currenIndex, s = s)
    }

    override fun onClick(view: View) {
        if (view === ivPrev) {
            if (isEnableAnimation) {
                LAnimationUtil.play(view = view, techniques = Techniques.Pulse)
            }
            currenIndex--
            updateUI()
        } else if (view === ivNext) {
            if (isEnableAnimation) {
                LAnimationUtil.play(view = view, techniques = Techniques.Pulse)
            }
            currenIndex++
            updateUI()
        }
    }

    fun getStringList(): ArrayList<String> {
        return stringList
    }

    fun getCurrenIndex(): Int {
        return currenIndex
    }
}
